### ISSUE TITLE
Abstract Objects Erroneously Reassembled?

### ISSUE THREAD

**Issue by protegesolutions** [obfuscar/obfuscar#186]
I have an API wrapper which uses **Newstonsoft.Json** to serialize API request parameters into string arguments.

I discovered that the arguments weren't being sent after obfuscating (all ok without).

Eg (real code)

        Dim oArgs As Object = New With {
            .filters = Filters,
            .offset = Offset,
            .count = Count,
            .sort_field = SortField,
            .sort_dir = SortDir,
            .fields = If(IncludeFields Is Nothing OrElse IncludeFields.Count = 0, Nothing, String.Join(",", IncludeFields)),
            .exclude_fields = If(IncludeFields IsNot Nothing AndAlso IncludeFields.Length > 0, Nothing, If(ExcludeFields Is Nothing OrElse ExcludeFields.Length = 0, Nothing, String.Join(",", ExcludeFields).Replace(sLinks, "")) + sLinks)}
        Dim sArgs As String = JsonConvert.SerializeObject(oArgs, New JsonSerializerSettings With {.NullValueHandling = NullValueHandling.Ignore, .MissingMemberHandling = MissingMemberHandling.Ignore})
 
In the last line above, the variable `sArgs` is empty ({}) though the object provided to JsonConvert is valid, eg:

`{ filters = , offset = 0, count = 100000, sort_field = date_created, sort_dir = desc, fields = lists.id,lists.name,lists.date_created,lists.stats,lists.modules,lists.stats.member_count, exclude_fields =  }`
    
Here's the (relevant parts of the) Obfuscar config:

    <Obfuscator>
      <Var name="InPath" value="." />
      <Var name="OutPath" value=".\ObfuscateTemp" />
      <Var name="HidePrivateApi" value="true" />
      <Var name="UseUnicodeNames" value="true" />
      <Var name="HideStrings" value="false" />
      <Var name="ReuseNames" value="false" />
      <Var name="KeyFile" value="..\..\protege.snk" />
      <AssemblySearchPath path="..\..\..\packages\Newtonsoft.Json.12.0.1\lib\net40" />
      <Module file="$(InPath)\Protege.ESP.MailChimp.API.dll">
        
      </Module>
    </Obfuscator>

I've tried all sorts of thing: setting `<SkipStringHiding type="Protege.ESP.MailChimp.API.Manager" name="*" />`, using code attribute `<Obfuscation(Exclude:=True)>` virtually everywhere, etc.

The only thing that resolves the issue is turning Obfuscar off.

**Comment by lextm** [comment]
Duplicate of #130.

**Comment by protegesolutions** [comment]
More on this ...

I traced it to something that happens to the abstract object `oArgs`. If I pass `JsonConvert.SerializeObject` a defined object, it works both straight and obfuscated.

However, even excluding the method, class, whatever from obfuscation causes something about the abstract object to be changed during processing. I created a very basic test method called `Test` so I could look at the IL and VB in both straight and obfuscated versions.

Here're the results:

_Unobfuscated VB_

    Public Shared Sub Test()
        Dim oArgs As Object = New With { .offset = 3, .sort_dir = "6" }
        Console.WriteLine(JsonConvert.SerializeObject(RuntimeHelpers.GetObjectValue(oArgs)))
    End Sub

_Unobfuscated  IL_

    .method public static void Test () cil managed 
    {
        .locals init (
            [0] object oArgs
        )

        IL_0000: nop
        IL_0001: ldc.i4.3
        IL_0002: ldstr "6"
        IL_0007: newobj instance void class VB$AnonymousType_0`2<int32, string>::.ctor(!0,  !1)
        IL_000c: stloc.0
        IL_000d: ldloc.0
        IL_000e: call object [mscorlib]System.Runtime.CompilerServices.RuntimeHelpers::GetObjectValue(object)
        IL_0013: call string [Newtonsoft.Json]Newtonsoft.Json.JsonConvert::SerializeObject(object)
        IL_0018: call void [mscorlib]System.Console::WriteLine(string)
        IL_001d: nop
        IL_001e: ret
    }

_Obfuscated VB_

    Public Shared Sub Test()
        Dim variable As Object = New  (Of Integer, String)(3, "6")
        Console.WriteLine(JsonConvert.SerializeObject(RuntimeHelpers.GetObjectValue(variable)))
    End Sub

_Obfuscated IL_

    .method public static void Test () cil managed 
    {
        .locals init (
            [0] object V_0
        )

        IL_0000: nop
        IL_0001: ldc.i4.3
        IL_0002: ldstr "6"
        IL_0007: newobj instance void class '\u00a0.\u00a0`2'<int32, string>::.ctor(!0,  !1)
        IL_000c: stloc.0
        IL_000d: ldloc.0
        IL_000e: call object [mscorlib]System.Runtime.CompilerServices.RuntimeHelpers::GetObjectValue(object)
        IL_0013: call string [Newtonsoft.Json]Newtonsoft.Json.JsonConvert::SerializeObject(object)
        IL_0018: call void [mscorlib]System.Console::WriteLine(string)
        IL_001d: nop
        IL_001e: ret
    }

The issue appears to be where I thought; the reassembly screws with the abstract object. It may  still be valid IL but `JsonConvert` sure doesn't like it.

So I'd guess that there is a small bug somewhere in **Obfuscar** that just needs abstract object handling fixed.

**Comment by protegesolutions** [comment]
> 
> 
> Duplicate of #130.

I don't think it is. I don't think the abstract object is being obfuscated, I think it being erroneous reassembled.
